### PR TITLE
ANP: Allow ANPs at the same priority and add events

### PR DESCRIPTION
**Pull request by tssurya**
#### What this PR does and why is it needed

See https://github.com/kubernetes-sigs/network-policy-api/issues/216 for details

#### Special notes for reviewers

We want to change the stance in OVNK about how we treated
ANPs at same priority. Moving forward we will allow multiple
ANPs at same priority and emit events for users.

Before this PR we were saying that the second ANP at the same
priority will not be allowed to be created, but now onwards that
won't be the case. - we are allowed to have 

#### How to verify it

Unit tests have been updated

### CODE REVIEW

**Comment by tssurya** [comment]
@trozet / @dceara : I hit the flake here: https://github.com/ovn-org/ovn-kubernetes/actions/runs/9317986672/job/25650454117?pr=4413 for that ICMP test on v4 ; wondering if the flake is because of the known bug or something else :) 

I'll open an u/s issue to track this if I see it happening again.

**Comment by tssurya** [comment]
@flavio-fernandes : PTAL

File: go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go [hunk 1]
```diff
@@ -105,6 +105,11 @@ func (c *Controller) ensureAdminNetworkPolicy(anp *anpapi.AdminNetworkPolicy) er
 	// supports upto 1000. The 0 (highest) corresponds to 30,000 in OVN world and
 	// 99 (lowest) corresponds to 20,100 in OVN world
 	if anp.Spec.Priority > ovnkSupportedPriorityUpperBound {
+		c.eventRecorder.Eventf(&v1.ObjectReference{
```
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go:108]: nice!

File: go-controller/pkg/ovn/controller/admin_network_policy/repair.go [hunk 2]
```diff
@@ -37,18 +35,6 @@ func (c *Controller) repairAdminNetworkPolicies() error {
 	existingANPs := map[string]*anpapi.AdminNetworkPolicy{}
 	for _, anp := range anps {
 		existingANPs[anp.Name] = anp
-		// let's populate the anpPriorityMap cache and add the correct ANP at the right priority
-		if anp.Spec.Priority > ovnkSupportedPriorityUpperBound {
```
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/controller/admin_network_policy/repair.go:41]: hm... this particular change should have been done in the other commit of this PR, right?

File: go-controller/pkg/ovn/admin_network_policy_test.go [hunk 3]
```diff
@@ -2066,4 +1976,124 @@ var _ = ginkgo.Describe("OVN ANP Operations", func() {
 			gomega.Expect(err).ToNot(gomega.HaveOccurred())
 		})
 	})
+	ginkgo.Context("Multiple ANPs at the same priority", func() {
```
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/admin_network_policy_test.go:1950]: Would it be a good idea to add a test case where 1) one of the 2 duplicate priorities are "not loaded" and then 2) the currently loaded priority is "unloaded / deleted" and 3) ensure that the unused admin netw policy is not used and enforced?
I also don't know if you had already tested cases when the duplication issue is resolved and the policy that was not chosen beginc taking effect. Just a question to ensure that is taken care of!

Comment (tssurya) [reply go-controller/pkg/ovn/admin_network_policy_test.go:1950]: oh so this was the old behaviour @flavio-fernandes where we were having two ANPs where we had a passive active mode, now what we plan to do is to allow to apply both the ANPs at the same priority. So moving forward both the policies will be chosen actually.

File: go-controller/pkg/ovn/controller/admin_network_policy/repair.go (hunk 2 above)
Comment (tssurya) [reply go-controller/pkg/ovn/controller/admin_network_policy/repair.go:41]: I can see why this confusion came from. So let me try to explain:

Before this PR:
1. We create ANP1 at prio5-> this gets created and stored in cache
2. ANP2 is created at prio5 -> we would look at the cache and say "oops I see another ANP at same prio" I won't "load" it or "create the OVN DB plumging for this" -> but cache contains ANP1 here
3. on upgrades we restart the controller so we loose the cache memory on this situation we'd need to on repair/startup ensure that we have loaded the older configured ANP1 in cache again (not ANP2) ==> In repair this is why we needed this full block

Now this behaviour is being changed by allowing the creation of two ANPs at the same priority

After this PR:
1. We create ANP1 at prio5 -> this gets created and stored in cache
2. we create ANP2 at prio5 -> we also create this actually fully including OVN DB plumbing -> but now we use the cache and see ANP1 is present and using that we emit an event ; so the cache now is only used for emitting events nothing else
3. on upgrades we restart, we don't care about populating the right one anymore so this entire block is removed from repair.

Does that make sense?

Comment (tssurya) [reply go-controller/pkg/ovn/controller/admin_network_policy/repair.go:41]: But... as I was explaining this to you I realised I don't have coverage for "deletion of ANP with that priority and ensuring no event is now emitted" I am going to add that now.

File: go-controller/pkg/ovn/admin_network_policy_test.go (hunk 3 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/admin_network_policy_test.go:1950]: yes, it does. Thanks!

File: go-controller/pkg/ovn/admin_network_policy_test.go [hunk 4]
```diff
@@ -2066,4 +1976,151 @@ var _ = ginkgo.Describe("OVN ANP Operations", func() {
...
+			anp, err := fakeOVN.fakeClient.ANPClient.PolicyV1alpha1().AdminNetworkPolicies().Create(context.TODO(), anp, metav1.CreateOptions{})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			gomega.Eventually(func() int {
+				anp, err = fakeOVN.fakeClient.ANPClient.PolicyV1alpha1().AdminNetworkPolicies().Get(context.TODO(), "harry-potter", metav1.GetOptions{})
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				return len(anp.Status.Conditions)
+			}).Should(gomega.Equal(1))
+			gomega.Expect(anp.Status.Conditions[0].Type).To(gomega.Equal("Ready-In-Zone-global"))
+			gomega.Expect(anp.Status.Conditions[0].Message).To(gomega.Equal("Setting up OVN DB plumbing was successful"))
+			gomega.Expect(anp.Status.Conditions[0].Reason).To(gomega.Equal("SetupSucceeded"))
```
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/admin_network_policy_test.go:1972]: Nice!

Comment (flavio-fernandes) [inline go-controller/pkg/ovn/admin_network_policy_test.go:1972]: Nice!

**Comment by flavio-fernandes** [comment]
Hit flake  https://github.com/ovn-org/ovn-kubernetes/issues/4144

https://github.com/ovn-org/ovn-kubernetes/actions/runs/9414617767/job/25934979357?pr=4413

```2024-06-07T10:01:26.5952221Z   [38;5;9m[FAILED][0m in [It] - /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/egressip.go:470 [38;5;243m@ 06/07/24 10:01:26.594[0m
2024-06-07T10:01:27.4332454Z   Jun  7 10:01:27.432: INFO: Checking the ovnkube-node and ovnkube-master (ovnkube-cluster-manager if interconnect=true) healthcheck ports in use
2024-06-07T10:01:27.4335681Z   Jun  7 10:01:27.432: INFO: Running '/usr/local/bin/kubectl --server=https://127.0.0.1:38421 --kubeconfig=/home/runner/ovn.conf --namespace=ovn-kubernetes get daemonset/ovnkube-node -o=jsonpath='{.spec.template.spec.containers[?(@.name=="ovnkube-controller")].env[?(@.name=="OVN_EGRESSIP_HEALTHCHECK_PORT")].value}''
```

**Review by flavio-fernandes — APPROVED** [review]
/lgtm

**Comment by tssurya** [comment]
the flake is unrelated:
```
Summarizing 1 Failure:
  [FAIL] e2e egress IP validation [OVN network] Using different methods to disable a node's availability for egress Should validate the egress IP functionality against remote hosts [It] disabling egress nodes impeding GRCP health check
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/egressip.go:470
```
https://github.com/ovn-org/ovn-kubernetes/actions/runs/9414617767/job/25934979357?pr=4413
merging this PR
